### PR TITLE
[NO MERGE] Uap

### PR DESCRIPTION
**Pull request by zabulus**
PR is just for review and sharing considerations about further implementation

### CODE REVIEW

File: global.json [hunk 1]
```diff
@@ -1,5 +1,10 @@
 {
   "projects": [
-    "src", "samples", "tests"
-  ]
+    "src",
+    "samples",
+    "tests"
+  ],
+  "sdk": {
+    "version": "1.0.0-preview2-1-003177"
```
Comment (adamsitnik) [inline global.json]: do you specify the sdk version explicity in order to get it running on VS 2017? If so, is the VS 2017 mandatory to compile the solution now? If yes, it's not a problem for me, I just want to know

File: samples/BenchmarkDotNet.Samples.Runner/project.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿{
+  "version": "1.0.0-*",
+  "buildOptions": {
+    "emitEntryPoint": true
+  },
+
+  "dependencies": {
+    "BenchmarkDotNet": "0.10.1-develop",
```
Comment (adamsitnik) [inline samples/BenchmarkDotNet.Samples.Runner/project.json]: you can make it a reference to the project, then there is no need to specify the version (which would require change with every update)

```json
"BenchmarkDotNet": {
          "target": "project"
        }
```

File: samples/BenchmarkDotNet.Samples/Program.cs [hunk 3]
```diff
@@ -1,9 +1,22 @@
 ﻿using System.Reflection;
 using BenchmarkDotNet.Running;
 
+#if UAP
+namespace System.Threading
+{
+    internal static class Thread
+    {
+        public static void Sleep(int milliseconds)
```
Comment (adamsitnik) [inline samples/BenchmarkDotNet.Samples/Program.cs]: I am not sure about the purpose of this class, but in general Thread.Sleep != Task.Delay so the name is wrong.

File: samples/BenchmarkDotNet.Samples/project.json [hunk 4]
```diff
@@ -33,22 +33,35 @@
...
+        "System.Threading.Tasks": "4.3.0"
       }
     }
   },
   "dependencies": {
     "BenchmarkDotNet": {
       "target": "project"
     }
+  },
+  "runtimes": {
```
Comment (adamsitnik) [inline samples/BenchmarkDotNet.Samples/project.json]: we can not support only hardcoded list of runtimes, but you probably know that

Is it possible to set the runtimes per target framework?

File: src/BenchmarkDotNet.Core/Attributes/Exporters/RPlotExporterAttribute.cs [hunk 5]
```diff
@@ -2,10 +2,12 @@
 
 namespace BenchmarkDotNet.Attributes.Exporters
 {
+#if !UAP
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Attributes/Exporters/RPlotExporterAttribute.cs]: please do that per whole file if possible (#if in the first line of the file)

File: src/BenchmarkDotNet.Core/Characteristics/CharacteristicHelper.cs [hunk 6]
```diff
@@ -56,12 +56,18 @@ public static IReadOnlyList<Characteristic> GetThisTypeCharacteristics(Type jobM
 
         private static IReadOnlyList<Characteristic> GetThisTypeCharacteristicsCore(Type jobModeType)
         {
-            var fieldValues = jobModeType.GetTypeInfo()
+            var fieldValues = jobModeType
+#if !UAP
+                .GetTypeInfo()
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Characteristics/CharacteristicHelper.cs]: I think that you could create an extension method for Type called GetTypeInfo() and get rid of all the ifdefines around this call

sth like:
```cs
#if UAP
internal static Type GetTypeInfo(this Type type) => type;
#endiff
```

File: src/BenchmarkDotNet.Core/Code/CodeGenerator.cs [hunk 7]
```diff
@@ -19,9 +19,14 @@ internal static class CodeGenerator
     {
         internal static string Generate(Benchmark benchmark)
         {
+            return Generate(benchmark, "BenchmarkProgram.txt");
+        }
+
+        internal static string Generate(Benchmark benchmark, string templateName)
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Code/CodeGenerator.cs]: can this method be private now?

File: src/BenchmarkDotNet.Core/Diagnosers/CompositeDiagnoser.cs [hunk 8]
```diff
@@ -19,13 +19,13 @@ public CompositeDiagnoser(params IDiagnoser[] diagnosers)
 
         public IColumnProvider GetColumnProvider() 
             => new CompositeColumnProvider(diagnosers.Select(d => d.GetColumnProvider()).ToArray());
-
+#if !UAP
         public void BeforeAnythingElse(Process process, Benchmark benchmark)
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Diagnosers/CompositeDiagnoser.cs]: does UAP has no definition of `Process` class?

File: src/BenchmarkDotNet.Core/Engines/RunResults.cs [hunk 9]
```diff
@@ -62,16 +62,18 @@ public IEnumerable<Measurement> GetMeasurements()
...
+        public void Print(Action<string> printAction)
+        {
             foreach (var measurement in GetMeasurements())
-                WriteLine(measurement.ToOutputLine());
+                printAction(measurement.ToOutputLine());
 
-            WriteLine(GCStats.WithTotalOperations(totalOperationsCount).ToOutputLine());
-            WriteLine();
+            printAction(GCStats.WithTotalOperations(totalOperationsCount).ToOutputLine());
+            printAction("\r\n");
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Engines/RunResults.cs]: can we use `Environment.NewLine` here?

File: src/BenchmarkDotNet.Core/Portability/RuntimeInformation.cs [hunk 10]
```diff
@@ -101,6 +101,8 @@ internal static string GetRuntimeVersion()
             }
 #if CLASSIC
             return $"Clr {System.Environment.Version}";
+#elif UAP
+            return "Uap 10.0";
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Portability/RuntimeInformation.cs]: is it possible to read the current version from somewhere?

File: src/BenchmarkDotNet.Core/Templates/UapBenchmarkProgram.txt [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+                    finally
+                    {
+						lc.Dispose();
+                        Application.Current.Exit();
+                    }
+                };
+
+                Task.Factory.StartNew(action);
+            }
+            catch (Exception ex)
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Templates/UapBenchmarkProgram.txt]: you started the task asynchronously, without awaiting it the catch will not work

btw why is the task needed? why not to run in in synchronous way?

File: src/BenchmarkDotNet.Core/Templates/UapBenchmarkProject.json [hunk 12]
```diff
@@ -0,0 +1,18 @@
+﻿{
+  "dependencies": {
+    "Microsoft.NETCore.UniversalWindowsPlatform": "5.2.2",
+    "system.console": "4.3.0",
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Templates/UapBenchmarkProject.json]: is dependency to system.console needed?

File: src/BenchmarkDotNet.Core/Toolchains/ConsoleHandler.cs [hunk 13]
```diff
@@ -50,7 +51,11 @@ private void HandlerCallback(object sender, ConsoleCancelEventArgs e)
                     return;
 
                 // Give it a bit of time to exit!
+#if UAP
+                System.Threading.Tasks.Task.Delay(500);
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Toolchains/ConsoleHandler.cs]: you are missing .Wait()

File: src/BenchmarkDotNet.Core/Toolchains/DotNetCli/DotNetCliGenerator.cs [hunk 14]
```diff
@@ -91,7 +96,11 @@ protected override void Cleanup(ArtifactsPaths artifactsPaths)
                 }
                 catch (Exception) when (attempt++ < 5)
                 {
+#if UAP
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Toolchains/DotNetCli/DotNetCliGenerator.cs]: does the `#if !UAP` in the top of the file excludes a chance for this `#if UAP`?

File: src/BenchmarkDotNet.Core/Toolchains/Uap/DevicePortalApiWrapper.cs [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+        public PackageStruct DeployApplication(string fullPath)
+        {
+            var installRequest = new RestRequest("api/app/packagemanager/package", Method.POST);
+            installRequest.AddCookie(csrfToken.Name, csrfToken.Value);
+            installRequest.AddCookie(wmid.Name, wmid.Value);
+            installRequest.AddFile(Path.GetFileName(fullPath), fullPath);
+            installRequest.AddQueryParameter("package", Path.GetFileName(fullPath));
+            installRequest.AddHeader("X-CSRF-Token", csrfToken.Value);
+            int i = 10;
+            for (;;)
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Toolchains/Uap/DevicePortalApiWrapper.cs]: I believe that `while(true)` would be easier to understand, or maybe we can simplify it further?

```cs
while (client.Execute(installRequest).StatusCode != HttpStatusCode.Accepted)
{
    if (--i == 0)
    {
         throw new InvalidOperationException("Install failed");
    }
}
```

File: src/BenchmarkDotNet.Core/Toolchains/Uap/UapToolchain.cs [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            var currentRuntime = "win10-x64";
+            if (!string.IsNullOrEmpty(currentRuntime))
+            {
+                return $"\"runtimes\": {{ \"{currentRuntime}\": {{ }} }},";
+            }
+
+            return string.Empty;
+        }
+    }
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Toolchains/Uap/UapToolchain.cs]: the private methods seems to not be used anywhere, can they be removed?

File: src/BenchmarkDotNet.Core/project.json [hunk 17]
```diff
@@ -48,7 +48,8 @@
         "System.Xml": "4.0.0.0"
       },
       "dependencies": {
-        "System.Threading.Tasks.Extensions": "4.0.0"
+        "System.Threading.Tasks.Extensions": "4.0.0",
+        "RestSharp": "105.2.3"
```
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/project.json]: "105.2.3" => this is a crazy version number ;)

File: global.json (hunk 1 above)
Comment (zabulus) [inline global.json]: I have some mess with cli tooling on my local PC. preview3 that I have installed, catched if no version specified, and it doesn't work.
I planned to revert this.

File: src/BenchmarkDotNet.Core/Characteristics/CharacteristicHelper.cs (hunk 6 above)
Comment (zabulus) [inline src/BenchmarkDotNet.Core/Characteristics/CharacteristicHelper.cs]: Nice, will fix this

File: global.json (hunk 1 above)
Comment (adamsitnik) [inline global.json]: @zabulus it's ok, I also needed this when installed VS 2017 on the same box.

File: src/BenchmarkDotNet.Core/Diagnosers/CompositeDiagnoser.cs (hunk 8 above)
Comment (zabulus) [inline src/BenchmarkDotNet.Core/Diagnosers/CompositeDiagnoser.cs]: There is System.Diagnostics.Process nuget package. But when I add ```runtimes``` section to a project where the package is added, tooling starts complaining about it.

File: src/BenchmarkDotNet.Core/Templates/UapBenchmarkProgram.txt (hunk 11 above)
Comment (zabulus) [inline src/BenchmarkDotNet.Core/Templates/UapBenchmarkProgram.txt]: It is UI application but without UI. 
I tried to run benchmark in constructor of Program, but no results appears. I assume that there is some timeout or restriction of operation that can be called in ctor.
Problem is that I can't reproduce that during device debug, only while benchmark running. So the fix is quite blind guess.

File: src/BenchmarkDotNet.Core/Templates/UapBenchmarkProject.json (hunk 12 above)
Comment (zabulus) [inline src/BenchmarkDotNet.Core/Templates/UapBenchmarkProject.json]: Will delete

File: src/BenchmarkDotNet.Core/Toolchains/DotNetCli/DotNetCliGenerator.cs (hunk 14 above)
Comment (zabulus) [inline src/BenchmarkDotNet.Core/Toolchains/DotNetCli/DotNetCliGenerator.cs]: I think whole Toolchains namespace should be excluded for UAP. There will be no support for running BDN as host.

File: src/BenchmarkDotNet.Core/Toolchains/Uap/DevicePortalApiWrapper.cs (hunk 15 above)
Comment (zabulus) [inline src/BenchmarkDotNet.Core/Toolchains/Uap/DevicePortalApiWrapper.cs]: Yep, that's much nicer

File: src/BenchmarkDotNet.Core/Toolchains/ConsoleHandler.cs (hunk 13 above)
Comment (zabulus) [inline src/BenchmarkDotNet.Core/Toolchains/ConsoleHandler.cs]: Whole file should excluded by UAP

File: src/BenchmarkDotNet.Core/Portability/RuntimeInformation.cs (hunk 10 above)
Comment (zabulus) [inline src/BenchmarkDotNet.Core/Portability/RuntimeInformation.cs]: Not sure what side is this code running at.

File: src/BenchmarkDotNet.Core/Templates/UapBenchmarkProgram.txt (hunk 11 above)
Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Templates/UapBenchmarkProgram.txt]: @zabulus this requires further investigation, we need to be sure that it does not affect results

Comment (zabulus) [inline src/BenchmarkDotNet.Core/Templates/UapBenchmarkProgram.txt]: Problem I've mentioned I've fixed with additional page (MainPage.xaml). The benchmark logic runs as background task, and application doesn't crash in release config.

What investigation should I perform to be sure that nothing affects a results?

Comment (adamsitnik) [inline src/BenchmarkDotNet.Core/Templates/UapBenchmarkProgram.txt]: @zabulus I am not 100% sure, but my guess is to check the difference between main app thread and the thread that task scheduler schedules for the task

sample problem could be: background thread vs foreground thread, different priorities 

@AndreyAkinshin @mattwarren @goldshtn what is your opinion on that?

**Pull request closed without merging**